xde/src/xde.rs

```diff
@@ -723,13 +854,42 @@ fn set_xde_underlay(req: &SetXdeUnderlayReq) -> Result<NoResp, OpteError> {
 
 const IOCTL_SZ: usize = core::mem::size_of::<OpteCmdIoctl>();
 
-static xde_ioc_list: [dld::dld_ioc_info_t; 1] = [dld::dld_ioc_info_t {
-    di_cmd: opte::api::XDE_DLD_OPTE_CMD as u32,
-    di_flags: dld::DLDCOPYINOUT,
-    di_argsize: IOCTL_SZ,
-    di_func: xde_dld_ioc_opte_cmd,
-    di_priv_func: secpolicy::secpolicy_dl_config,
-}];
+#[no_mangle]
+unsafe extern "C" fn xde_getinfo(
+    dip: *mut dev_info,
+    cmd: ddi_info_cmd_t,
+    arg: *mut c_void,
+    resultp: *mut *mut c_void,
+) -> c_int {
+    if xde_dip.is_null() {
+        return DDI_FAILURE;
+    }
+
+    // If this isn't one of our private minors,
+    // let the GLDv3 framework handle it.
+    let minor = getminor(arg as dev_t);
+    if minor < mac_private_minor() {
+        return dld_getinfo(dip, cmd, arg, resultp);
+    }
+
+    // We currently only expose a single minor node,
+    // bail on anything else.
+    if minor != XDE_CTL_MINOR {
+        warn!("unexpected minor number {minor}");
+        return DDI_FAILURE;
+    }
```

Comment:
Isn't `arg` a `dev_t` in either case? From `getinfo(9e)`:
```
       When cmd is set to DDI_INFO_DEVT2DEVINFO, getinfo() should return the
       dev_info_t pointer associated with the dev_t arg.  The dev_info_t
       pointer should be returned in the field pointed to by resultp.


       When cmd is set to DDI_INFO_DEVT2INSTANCE, getinfo() should return the
       instance number associated with the dev_t arg. The instance number
       should be returned in the field pointed to by resultp.
```

I guess in the limit there could be a new variant of `cmd` that isn't. I can rewrite this to be a little more resilient in that case. I'll probably update `ddi_info_cmd_t` to not be a Rust enum then lest we end up with weird codegen from rustc assuming that's exhaustive.

Reply:
Yes, it is a `dev_t` in the case of either command that is documented today, but the type is dependent on command and checking that first. We ask folks to write these (and it happens to varying degrees) such that another command can be added here. Because `getmajor()` and `getminor()` are just operating on integers, treating the raw pointer as that isn't the worst in terms of things that can go wrong, but I think it's important to try to phrase this such that we follow the intent.